datafusion/sql/src/utils.rs

```diff
@@ -246,3 +246,168 @@ pub(crate) fn normalize_ident(id: Ident) -> String {
+    unescape_unicode::<8>(queue)
+}
+
+fn unescape_unicode<const NUM: usize>(queue: &mut VecDeque<char>) -> Option<char> {
+    let mut s = String::new();
+    for _ in 0..NUM {
+        s.push(queue.pop_front()?);
+    }
+    to_char::<16>(&s)
+}
+
+// Octal byte value. \o, \oo, \ooo (o = 0–7)
+fn unescape_octal(c: char, queue: &mut VecDeque<char>) -> Option<char> {
+    let mut s = String::new();
+
+    match c {
+        '0' | '1' | '2' | '3' => {
+            s.push(c);
+            for _ in 0..2 {
+                match next_octal_digest(queue) {
+                    Some(c) => s.push(c),
+                    None => break,
+                }
+            }
+        }
+        '4' | '5' | '6' | '7' => {
+            s.push(c);
+            if let Some(c) = next_octal_digest(queue) {
+                s.push(c);
+            }
+        }
+        _ => return None,
+    }
+    to_char::<8>(&s)
+}
+
+#[inline]
+fn next_octal_digest(queue: &mut VecDeque<char>) -> Option<char> {
+    match queue.front() {
+        Some(c) if c.is_digit(8) => queue.pop_front(),
+        _ => None,
+    }
+}
+
+#[inline]
+fn to_char<const RADIX: u32>(s: &str) -> Option<char> {
+    match u32::from_str_radix(s, RADIX) {
+        Err(_) => None,
+        Ok(n) => char::from_u32(n),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::unescape;
+
+    fn check_unescape(s: &str, expected: Option<&str>) {
+        assert_eq!(unescape(s), expected.map(|s| s.to_string()));
+    }
+
+    #[test]
+    fn test_unescape() {
+        check_unescape(r"\b", Some("\u{0008}"));
+        check_unescape(r"\f", Some("\u{000C}"));
+        check_unescape(r"\t", Some("\t"));
+        check_unescape(r"\r\n", Some("\r\n"));
+        check_unescape(r"\/", Some("/"));
+        check_unescape(r"/", Some("/"));
+        check_unescape(r"\\", Some("\\"));
+
+        // 16 and 32-bit hexadecimal Unicode character value
+        check_unescape(r"\u4c91", Some("\u{4c91}"));
+        check_unescape(r"\u4c916", Some("\u{4c91}6"));
+        check_unescape(r"\U0010FFFF", Some("\u{10FFFF}"));
+        check_unescape(r"\u", None);
+        check_unescape(r"\U", None);
+        check_unescape(r"\U1010FFFF", None);
+
+        // hexadecimal byte value
+        check_unescape(r"\xCAD", Some("\u{00ca}D"));
+        check_unescape(r"\xA9", Some("\u{00a9}"));
+        check_unescape(r"\x4B", Some("\u{004b}"));
+        check_unescape(r"\x4", Some("\u{0004}"));
+        check_unescape(r"\x4L", Some("\u{0004}L"));
+        check_unescape(r"\x", Some("x"));
+        check_unescape(r"\xP", Some("xP"));
+
+        // octal byte value
+        check_unescape(r"\1", Some("\u{0001}"));
+        check_unescape(r"\12", Some("\u{000a}"));
+        check_unescape(r"\123", Some("\u{0053}"));
+        check_unescape(r"\1232", Some("\u{0053}2"));
+        check_unescape(r"\4", Some("\u{0004}"));
+        check_unescape(r"\45", Some("\u{0025}"));
+        check_unescape(r"\450", Some("\u{0025}0"));
```

Comment:
Was checking against Postgres, got some interesting behaviour:

```
postgres=# select e'\450' as a;
 a
---
 (
(1 row)
```

Also can you add these test cases as I'm curious as to how this implementation will handle it vs Postgres

```
postgres=# select e'\077' as a;
 a
---
 ?
(1 row)

postgres=# select e'\078' as a;
   a
-------
 \x078
(1 row)

postgres=# select e'\079' as a;
   a
-------
 \x079
(1 row)

postgres=# select e'\080' as a;
ERROR:  invalid byte sequence for encoding "UTF8": 0x00
```

Reply:
When I wrote this `unescape` function, I was referring to the behavior of pg.

Reply:
> When I wrote this `unescape` function, I was referring to the behavior of pg.

But it may be incomplete 😅